File: 2018/Day4.hs
module Day4 where

import Data.Function
import Data.List
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as M
import Data.Maybe
import Data.Vector (Vector)
import qualified Data.Vector as V

data Record = BeginsShift {guardId :: Int}
            | FallsAsleep {minute :: Int}
            | WakesUp     {minute :: Int}
            deriving (Show)

parseRecord :: [Char] -> Record
parseRecord rstring
    | "falls asleep" `isInfixOf` rstring = FallsAsleep minute
    | "wakes up" `isInfixOf` rstring     = WakesUp minute
    | otherwise                          = BeginsShift guardId
    where minute  = read $ take 2 $ drop 15 rstring
          guardId = read $ takeWhile ((/=) ' ') $ drop 1 $ dropWhile ((/=) '#') rstring

updateMinuteCount :: (Num a) => Int -> Int -> Vector a -> Vector a
updateMinuteCount start end minuteCount = V.map update $ V.zip indices minuteCount where
    update (i, m) = if start <= i && i < end then m + 1 else m
    indices       = V.generate (V.length minuteCount) id

buildSleepLog :: Int -> Maybe Int -> Map Int (Int, Vector Int) -> [Record] -> Map Int (Int, Vector Int)
buildSleepLog _ _ rmap []                     = rmap
buildSleepLog guardId asleepSince rmap (r:rs) = case r of BeginsShift newGuardId -> buildSleepLog newGuardId Nothing rmap rs
                                                          FallsAsleep minute     -> buildSleepLog guardId (Just minute) rmap rs
                                                          WakesUp minute         -> buildSleepLog guardId Nothing (M.insert guardId (updatedLog minute) rmap) rs
    where minutesAsleep currentMinute = currentMinute - (fromJust asleepSince)
          updatedLog currentMinute    = case M.lookup guardId rmap of Nothing                -> (minutesAsleep currentMinute, updateMinuteCount (fromJust asleepSince) currentMinute $ V.replicate 60 0)
                                                                      Just (total, minutes)  -> (minutesAsleep currentMinute + total, updateMinuteCount (fromJust asleepSince) currentMinute minutes)

maximumIndex :: (Ord a) => Vector a -> Int
maximumIndex xs = let maxValue = V.maximum xs in fromJust $ V.findIndex ((==) maxValue) xs

main :: IO ()
main = do
    records <- map parseRecord . sort . lines <$> readFile "2018/sampleinputs/day4.txt"
    let sleepLog = buildSleepLog 0 Nothing M.empty records
    let (sleepingBeauty1, log1) = head $ sortBy (flip compare `on` (fst . snd)) $ M.toList sleepLog
    let (sleepingBeauty2, log2) = head $ sortBy (flip compare `on` (maximum . snd . snd)) $ M.toList sleepLog
    let part1 = sleepingBeauty1 * (maximumIndex $ snd log1)
    let part2 = sleepingBeauty2 * (maximumIndex $ snd log2)
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2

File: 2018/day4.py
from datetime import datetime
from enum import Enum, auto
import re

guard_id_regex = re.compile(r'#\d+')


class Record():
    def __init__(self, record_string):
        self.timestamp = datetime.strptime(record_string[1:17], '%Y-%m-%d %H:%M')
        guard_id = guard_id_regex.search(record_string)
        self.guard_id = guard_id[0][1:] if guard_id else None
        if 'wakes up' in record_string:
            self.type = 'WAKES_UP'
        elif 'falls asleep' in record_string:
            self.type = 'FALLS_ASLEEP'
        else:
            self.type = 'BEGINS_SHIFT'


def track_sleep_minutes(minutes_already_slept, start_minute, end_minute):
    for minute in range(start_minute, end_minute):
        minutes_already_slept[minute] += 1


def build_sleep_log(records):
    sleep_log = {}
    current_guard_id = records[0].guard_id
    asleep_since = None
    for record in records:
        if record.type == 'BEGINS_SHIFT':
            current_guard_id = record.guard_id
            asleep_since = None
        if record.type == 'FALLS_ASLEEP':
            asleep_since = record.timestamp
        if record.type == 'WAKES_UP':
            time_asleep = record.timestamp - asleep_since
            if current_guard_id in sleep_log:
                sleep_log[current_guard_id]['time_asleep'] += time_asleep
            else:
                sleep_log[current_guard_id] = {'time_asleep': time_asleep, 'minutes': [0] * 60}
            for minute in range(asleep_since.minute, record.timestamp.minute):
                sleep_log[current_guard_id]['minutes'][minute] += 1

    return sleep_log


if __name__ == '__main__':
    with open('2018/sampleinputs/day4.txt') as file:
        records = sorted([Record(r) for r in file.read().split('\n')[:-1]], key=lambda r: r.timestamp)

        sleep_log = build_sleep_log(records)

        sleeping_beauty1, log1 = sorted(sleep_log.items(), key=lambda x: x[1]['time_asleep'], reverse=True)[0]
        part1 = int(sleeping_beauty1) * int(log1['minutes'].index(max(log1['minutes'])))
        print('Part 1: {}'.format(part1))

        sleeping_beauty2, log2 = sorted(sleep_log.items(), key=lambda x: max(x[1]['minutes']), reverse=True)[0]
        part2 = int(sleeping_beauty2) * int(log2['minutes'].index(max(log2['minutes'])))
        print('Part 2: {}'.format(part2))
